assets/src/components/block-preview-label.js

```diff
@@ -2,41 +2,79 @@
  * WordPress dependencies
  */
 import { Fragment } from '@wordpress/element';
+import { getBlockType } from '@wordpress/blocks';
 import { BlockIcon } from '@wordpress/block-editor';
+import { withSelect } from '@wordpress/data';
+import { __experimentalGetSettings as getDateSettings, dateI18n } from '@wordpress/date';
+import { __ } from '@wordpress/i18n';
 
-export default ( { label: name, block, blockType, displayIcon = true, alignIcon = 'left', accessibilityText } ) => {
-	if ( ! block ) {
-		return name;
-	}
+const BlockPreviewLabel = ( { content, icon, displayIcon = true, alignIcon = 'left', accessibilityText } ) => {
+	return (
+		<Fragment>
+			{ displayIcon && 'left' === alignIcon && <BlockIcon icon={ icon } /> }
+			{ content.length > 20 ? `${ content.substr( 0, 20 ) }…` : content }
+			{ accessibilityText && <span className="screen-reader-text">{ accessibilityText }</span> }
+			{ displayIcon && 'right' === alignIcon && <BlockIcon icon={ icon } /> }
+		</Fragment>
+	);
+};
+
+export default withSelect( ( select, { block } ) => {
+	const { getEditedPostAttribute } = select( 'core/editor' );
+	const { getAuthors, getMedia } = select( 'core' );
+
+	const blockType = getBlockType( block.name );
 
 	let label = blockType.title;
 	let content;
 
-	// Todo: Cover more special cases if needed.
 	switch ( block.name ) {
 		case 'core/image':
 			if ( block.attributes.url ) {
-				content = block.attributes.url.slice( block.attributes.url.lastIndexOf( '/' ) ).slice( 0, 30 );
+				content = block.attributes.url.slice( block.attributes.url.lastIndexOf( '/' ) ).slice( 1, 30 );
 
 				if ( content.length > 0 ) {
 					label = content;
 				}
 			}
 
+			if ( block.attributes.id ) {
+				const media = getMedia( block.attributes.id );
+
+				if ( media ) {
+					label = media.caption.raw || media.title.raw || label;
+				}
```

Comment:
Perhaps we could change the order to use the title if available, otherwise use the caption, and use the id as a baseline.

Reply:
IDs are never exposed to the user in WordPress, so I‘d leave that one out.